src/telemetry/sensor/mod.rs

```diff
@@ -36,27 +41,29 @@ impl Handle {
     where
         F: FnOnce() -> event::Event,
     {
-        if let Some(tx) = self.0.as_mut() {
-            // We may want to capture timestamps here instead of on the consumer-side...  That
-            // level of precision doesn't necessarily seem worth it yet.
-
+        if let Some(inner) = self.0.as_mut() {
             let ev = mk();
             trace!("event: {:?}", ev);
 
-            if tx.lossy_send(ev).is_err() {
-                debug!("dropped event");
+            if let Ok(mut taps) = inner.taps.lock() {
```

Comment:
Taking this lock for every event is a little worrisome. It seems benchmarks weren't hurt, though they also don't run while a user is tapping, right?

Reply:
the lock is acquired regardless of whether taps are active, but it will be held for the full tap inspection -- so yes, tap can impact latency. We can likely change this to use a ReadWriteLock pretty easily, and, ultimately i want to do away with events entirely so that lock acquisition occurs at most once per request

Reply:
Right, but since no taps are active, there's never contention for the lock in those benchmarks. Eh, it's *probably* fine.

Reply:
aiui there's only contention while the tap request is being registered, not throughout the full lifetime of the  tap observation -- basically a map insert https://github.com/linkerd/linkerd2-proxy/blob/7788f60e0ef92b51f69828be689683db22f964b9/src/control/observe.rs#L75-L87